ddl/table.go

```diff
@@ -735,7 +736,7 @@ func onModifyTableCharsetAndCollate(t *meta.Meta, job *model.Job) (ver int64, _
 	tblInfo.Collate = toCollate
 	// update column charset.
 	for _, col := range tblInfo.Columns {
-		if typesNeedCharset(col.Tp) {
+		if field_types.HasCharset(&col.FieldType) {
```

Comment:
I guess `typesNeedCharset` can be replaced by `HasCharset`. Another code that calls this function is:

https://github.com/pingcap/tidb/blob/5c21fa4b4345ef61d8ed936c86d3ee477fcf24f6/ddl/ddl_api.go#L299

Can you confirm it here?

Reply:
No, https://github.com/pingcap/tidb/blob/5c21fa4b4345ef61d8ed936c86d3ee477fcf24f6/ddl/ddl_api.go#L299 here may be have not set the column Flag, use `HasCharset ` here may be got some error.

Reply:
so why you use `HashCharset` finally?

Reply:
`HashCharset ` can be used here(`ddl/table.go #739`), but `ddl/ddl_api.go #299` should use `typesNeedCharset` function.

Reply:
Sorry, I still don't quite understand, why is it correct to change to `HasCharset`?